PSQLADM-321: Initial implementation of admin script for golang based scheduler

venkatesh-prasad-v opened the pull request:
https://jira.percona.com/browse/PSQLADM-321
https://jira.percona.com/browse/PSQLADM-330

Test result: https://gist.github.com/venkatesh-prasad-v/30d0eef27a1e7adac45561640c72481a

kennt-percona committed:
go scheduler initial implementation

Commit:
1. Fixed script error due to missing `fi` 2. Moved the check for `my_print_defaults` to a later stage in order to make it work with `--help`

Commit:
Update percona-scheduler submodule pointer

Commit:
Print scheduler version in scheduler bats script

Commit:
The following problems have been fixed

1. --update-mysql-version fails with unbound variable error.

2. --version-check test fails with version mismatch.

3. logFile in tests/testsuite.toml pointed to an inexistent file.

Commit:
Copy config files to workdir during setup_workdir.sh

Commit:
Add support for MariaDB 10.6

Commit:
Avoid duplicated test names in generic-test.bats

Commit:
The following improvements were done

1. PXC_START_TIMEOUT increased to 60 seconds.
2. Make the script to exit on timeout. Earlier it ignored the timeout
   error.
3. The script gracefully shuts down the proxysql and mysqld processes.
   Earlier it used SIGKILL to terminate the processes.
4. Proxysql is now started with `--foreground` option to enable log
   capturing.
5. lockfilepath in tests/testsuite.toml pointed to an invalid file. This
   is now fixed.

Commit:
Initial implementation of admin script for golang based scheduler

Merge remote-tracking branch 'kenn/go_scheduler' into v2

Commit:
Update README.md from the latest outut of `proxysql-admin --help`

Commit:
Add usage details in README.md

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/signed)](https://cla-assistant.percona.com:443/percona/proxysql-admin-tool?pullRequest=217) <br/>All committers have signed the CLA.

Tusamarco commented:
Ok will check it as soon as I am back next  Monday.

On Tue, Feb 15, 2022, 8:42 AM kamil-holubicki ***@***.***>
wrote:

> @kamil-holubicki <https://github.com/kamil-holubicki> requested your
> review on: #217 <https://github.com/percona/proxysql-admin-tool/pull/217>
> Initial implementation of admin script for golang based scheduler.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/percona/proxysql-admin-tool/pull/217#event-6072665923>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKOQJ3C2LRMX5FX4TZIA2TU3H7XHANCNFSM5ONE57EQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

Commit:
Modified the script to assign higher weightage to both writer hostgroup
and writer config hostgroup when node specified in the `--write-node`
option when used along with `--update-cluster` option is passed.

mohitj1988 commented on build_scheduler.sh:
```diff
@@ -0,0 +1,27 @@
+
+cd percona-scheduler
+
+go build -a -ldflags "-X main.pxcSchedulerHandlerVersion=${PROXYSQL_ADMIN_VERSION}" -o pxc_scheduler_handler
```
We need to modify the scripts to display errors in case compilation fails during go build stage

venkatesh-prasad-v replied:
Fixed.

mohitj1988 commented on tests/setup_workdir.sh:
```diff
@@ -114,12 +114,17 @@ fi
+for file in proxysql-admin proxysql-common proxysql-admin-common percona-scheduler-admin proxysql-login-file
 do
   ln -s $PROXYSQL_ADMIN_BASEDIR/$file $WORKDIR
 done;
```
As discussed, also copy the pxc_scheduler_handler binaries into workdir

venkatesh-prasad-v replied:
Fixed.

Commit:
Following improvements have been done.

1. Updated README to include mysql client in the prerequisites section.
2. Added checks to bail out the errors when
   - golang is not installed
   - go build process failed with errors.
3. Use `command -v` instead of `which` to determine the binary path.
4. Create a soft-link to the pxc_scheduler_handler binary when setting up the
   workdir.

Commit:
PSQLADM-364: fix typo errors in testsuite.toml file

https://jira.percona.com/browse/PSQLADM-364

Commit:
PSQLADM-362: Scheduler admin does not display correct result when --update-cluster and --enable are used together
PSQLADM-363: Using --update-cluster before --enable does not give correct result with scheduler script

https://jira.percona.com/browse/PSQLADM-362
https://jira.percona.com/browse/PSQLADM-363

Changes:
- In singlewrite mode, scheduler Admin script shall now wait until the nodes are
  processed by the scheduler when --enable option is passed (even when
  passed along with --update-cluster).

- Passing --update-cluster option without configuring it shall result in
  below error

    ==== OUTPUT BEGIN ====

    $ ercona-scheduler-admin --config-file=config.toml --update-cluster
    ERROR (line:2373) : The cluster (with writer hostgroup: 100) has not been configured in ProxySQL

    ==== OUTPUT END ====

Commit:
PSQLADM-366: Scheduler admin script allows invalid options

https://jira.percona.com/browse/PSQLADM-366

- Added the error message when WriterIsAlsoReader=0 when maxNumWriters >
  1 as the options are disallowed by the scheduler itself.

Additional changes
------------------

- Setting maxNumWriters=1 shall always enable the cluster in singlewrite
  mode implicitly.

- Replaced "lineno" with "LINENO" to print the line number during the
  error.

- writers_are_readers variable checks are done in boolean way using 0s
  and 1s, instead of text (using "yes" / "no")

Commit:
Implement options for Automatic Weight Assignment and Node weight
Updation

Summary: Added 3 new options
1. --auto-assign-weights to assign the weights automatically when in
   singlewrite mode.
2. --update-read-weight<IP:PORT,WT> When used with --update-cluster, this
   option shall assign the specified read weight to the given node.
2. --update-write-weight<IP:PORT,WT> When used with --update-cluster, this
   option shall assign the specified write weight to the given node.

Additional changes:
- Updated README
- Updated the submodule pointer
- Added BATS tests
